File: server/client/src/main/kotlin/cz/uhk/mois/client/domain/ServiceCase.kt
package cz.uhk.mois.client.domain

import cz.uhk.mois.client.controller.model.ServiceCaseType
import org.springframework.data.annotation.Id
import org.springframework.data.relational.core.mapping.Table
import java.time.Instant
import javax.persistence.EnumType
import javax.persistence.Enumerated
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType

@Table(name = "SERVICE_CASES")
class ServiceCase(
    @Id
    @GeneratedValue(strategy = GenerationType.TABLE)
    var id: Long? = null,

    @Enumerated(EnumType.STRING)
    var type: ServiceCaseType,
    var serialNumber: String,
    val message: String,
    val name: String,
    val surname: String,
    val email: String,
    val phone: String?,
    val street: String?,
    val houseNumber: Int?,
    val city: String?,
    val postalCode: String?,
    val dateBegin: Instant = Instant.now(),
    val dateEnd: Instant?
)

File: web/src/api/models/ServiceCase.ts
import { SelectField } from '../../common/forms/SelectField'
import { Device, EMPTY_DEVICE } from './Device'
import { ServiceCaseMessage } from './ServiceCaseMessage'
import { EMPTY_USER, User } from './User'

export const ServiceCaseListSorting: SelectField[] = [
    { code: 'date-desc', value: 'Nejnovější' },
    { code: 'date-asc', value: 'Nestarší' },
]

export type ServiceCaseList = {
    hasNext: boolean
    hasPrev: boolean
    data: ServiceCaseListItem[]
    page: number
    totalPages: number
}

export type ServiceCaseListItem = {
    id: number
    dateBegin: string
    dateEnd: string
    client: string
    message: string
    newMessagesCount: number
    stateId: number
    operators: string[]
}

export const EMPTY_SERVICE_CASE_LIST: ServiceCaseList = {
    hasNext: false,
    hasPrev: false,
    page: 1,
    totalPages: 1,
    data: [],
}

export type ServiceCaseForm = {
    deviceTypeId: string
    caseTypeId: string
    serialNumber: string
    message: string

    name: string
    surname: string
    email: string
    phone: string

    street: string
    houseNumber: string
    city: string
    postalCode: string
}

export type ServiceCaseCreated = {
    id: string
    hash: string
    email?: string
}

export const EMPTY_SERVICE_CASE_CREATED: ServiceCaseCreated = {
    id: '',
    hash: '',
    email: '',
}

export type ServiceCaseDetail = {
    serviceCase: {
        id: string
        stateId: string
        caseTypeId: string
        dateBegin: string
        dateEnd: string
        hash: string
    }
    operators: User[]
    client: User
    device: Device
    messages: ServiceCaseMessage[]
}

export const EMPTY_SERVICE_CASE_DETAIL: ServiceCaseDetail = {
    serviceCase: {
        id: '',
        stateId: '',
        caseTypeId: '',
        dateBegin: '',
        dateEnd: '',
        hash: '',
    },
    operators: [],
    client: EMPTY_USER,
    device: EMPTY_DEVICE,
    messages: [],
}

export type ServiceCaseType = {
    code: string
    value: string
}

export type ServiceCaseState = {
    code: string
    value: string
}
